tests/testsuite/update.rs

```diff
@@ -2064,3 +2112,508 @@ fn update_breaking_specific_packages() {
+[UPDATING] `[..]` index
+[UPGRADING] mixed-compatibility ^1.0 -> ^2.0
+[LOCKING] 1 package to latest compatible version
+[ADDING] mixed-compatibility v2.0.1
+
+"#]])
+        .run();
+}
+
+#[cargo_test]
+fn update_breaking_mixed_pinning_renaming() {
+    Package::new("mixed-pinned", "1.0.0").publish();
+    Package::new("mixed-ws-pinned", "1.0.0").publish();
+    Package::new("renamed-from", "1.0.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [workspace]
+                members = ["foo", "bar", "baz"]
+
+                [workspace.dependencies]
+                mixed-ws-pinned = "=1.0"
+            "#,
+        )
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                mixed-pinned = "=1.0"
+                mixed-ws-pinned.workspace = true
+                renamed-to = { package = "renamed-from", version = "1.0" }
+            "#,
+        )
+        .file("foo/src/lib.rs", "")
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [package]
+                name = "bar"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                mixed-pinned = "1.0"
+                mixed-ws-pinned = "1.0"
+                renamed-from = "1.0"
+            "#,
+        )
+        .file("bar/src/lib.rs", "")
+        .file(
+            "baz/Cargo.toml",
+            r#"
+                [package]
+                name = "baz"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [target.'cfg(windows)'.dependencies]
+                mixed-pinned = "1.0"
+
+                [target.'cfg(unix)'.dependencies]
+                mixed-pinned = "=1.0"
+            "#,
+        )
+        .file("baz/src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile").run();
+
+    Package::new("mixed-pinned", "2.0.0").publish();
+    Package::new("mixed-ws-pinned", "2.0.0").publish();
+    Package::new("renamed-from", "2.0.0").publish();
+
+    p.cargo("update -Zunstable-options --breaking")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr_data(str![[r#"
+[UPDATING] `[..]` index
+[UPGRADING] mixed-pinned ^1.0 -> ^2.0
+[UPGRADING] mixed-ws-pinned ^1.0 -> ^2.0
+[UPGRADING] renamed-from ^1.0 -> ^2.0
+[LOCKING] 3 packages to latest compatible versions
+[ADDING] mixed-pinned v2.0.0
+[ADDING] mixed-ws-pinned v2.0.0
+[ADDING] renamed-from v2.0.0
+
+"#]])
+        .run();
+
+    let root_manifest = p.read_file("Cargo.toml");
+    let foo_manifest = p.read_file("foo/Cargo.toml");
+    let bar_manifest = p.read_file("bar/Cargo.toml");
+    let baz_manifest = p.read_file("baz/Cargo.toml");
```

Comment:
Would it work to give these workspace members names tied to their role in the test?  For example, they can be "pinned", "unpinned", "mixed-pinned".  That'd make it easier to follow what the test is doing

Reply:
Fixed.